### PR TITLE
feat: Check if signing fails when revoked key is used

### PR DESCRIPTION
**Pull request by ukstv**

### CODE REVIEW

**Comment by ukstv** [comment]
Hey @stbrody It is draft, because I am really not sure if the code is all right, and fits the whole test suite.

**Review by oed — COMMENTED** [review]
Looks good so far!

File: src/__tests__/key_revocation.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    const firstSigner = threeIdProvider.keychain._keyring.getSigner();
+
+    // 2. Create tile
+    const tile = await TileDocument.create(ceramicClient, {
+        stage: "Signed by vanilla",
+    });
+    await tile.update({ stage: "Signed second time" }, undefined, {
+        anchor: true,
+    });
+    await waitForAnchor(tile, 60*60);
```
Comment (stbrody) [inline src/__tests__/key_revocation.ts:71]: this means waiting an hour before the test fails if the anchor batch fails, which seems very long.  The other tests use 15 minutes, probably makes sense to be consistent with them.  Might even be nice to pull ANCHOR_TIMEOUT to be a re-usable constant that the tests share

File: src/__tests__/key_revocation.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        authSecret: seed,
+        authId: "first",
+        ceramic: ceramicClient,
+    });
+    const did = new DID({
+        provider: threeIdProvider.getDidProvider(),
+        resolver: resolver,
+    });
+    await did.authenticate();
+    ceramicClient.did = did;
```
Comment (stbrody) [inline src/__tests__/key_revocation.ts:60]: If you're messing with the ceramic's DID in this test it would probably be good to restore it to its original state at the end so that you don't interfere with other tests

File: src/__tests__/key_revocation.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import ThreeIdResolver from "@ceramicnetwork/3id-did-resolver";
+import { Resolver } from "did-resolver";
+import {DID, GeneralJWS} from "dids";
+import ThreeIdProvider from "3id-did-provider";
+import {TileDocument} from "@ceramicnetwork/stream-tile";
+import {waitForAnchor} from "../utils";
+import * as didJWT from "did-jwt";
+
+declare global {
+    const ceramic: CeramicApi;
```
Comment (stbrody) [inline src/__tests__/key_revocation.ts:18]: If you're only using one it should be `ceramic` not `ceramicClient`.  `ceramicClient` is *always* an http client whereas `ceramic` is sometimes an http client and sometimes a core node depending on the environment, so you get more coverage for free by using `ceramic`.

**Review by stbrody — COMMENTED** [review]
Overall looks good, a few small suggestions inline.

I guess the reason for having this as an integration test instead of a test in js-ceramic is just to have a real CAS rather than an InMemoryAnchorService?  Might be nice to have a version of this test in the js-ceramic repo with an InMemoryAnchorService also as it'll run much faster and give us faster feedback if something breaks before anything gets committed or released in ceramic.

**Comment by ukstv** [comment]
@stbrody addressed all the comments.

**Review by oed — CHANGES_REQUESTED** [review]
I would just add one more check for future safety.

File: src/__tests__/key_revocation.ts (hunk 1 above)
Comment (ukstv) [reply src/__tests__/key_revocation.ts]: done

File: src/__tests__/key_revocation.ts (hunk 3 above)
Comment (ukstv) [reply src/__tests__/key_revocation.ts:18]: Why it surprises you @oed ? It is the same interface for core and client.

Comment (ukstv) [reply src/__tests__/key_revocation.ts:18]: Now the code uses `ceramic`.

File: src/__tests__/key_revocation.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        const compactJWS = await didJWT.createJWS(payload, firstSigner, {
+            kid: firstKid,
+        });
+        return toGeneralJWS(compactJWS);
+    };
+
+    // 4. This should blow
+    await expect(tile.update({ stage: "Third round" }, undefined, {
+        anchor: true,
+    })).rejects.toThrow(/JWS was signed with a revoked DID version/)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/__tests__/key_revocation.ts]: kk. added.

File: src/__tests__/key_revocation.ts (hunk 2 above)
Comment (ukstv) [reply src/__tests__/key_revocation.ts]: done.

File: src/__tests__/key_revocation.ts (hunk 3 above)
Comment (oed) [reply src/__tests__/key_revocation.ts:18]: I was surprised by Spencer's comment :)

**Review by oed — APPROVED** [review]

**Review by stbrody — COMMENTED** [review]
this PR LGTM, though I wonder if now we should expand this new test with some of the edge cases that have come up in the review cycle for the changes being tested.  Like the "lone genesis" case for instance.

**Comment by oed** [comment]
That's a good point @stbrody.
@ukstv Can you make sure we capture a story for the test Spencer suggested (no need to implement in this sprint).

**Comment by ukstv** [comment]
This is captured in https://github.com/3box/ceramic-integration-tests/issues/62
